drivers: memc: add driver for stm32 ospi psram

mariopaja opened the pull request:
## Add  STM32 OSPI PSRAM driver support in memory mapped mode

Current driver configuration based on `aps6408l` on **STM32U5xx** series

## Test application 1
https://github.com/mariopaja/hello_psram


**Custom U575 Board:**
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       23572 B         2 MB      1.12%
             RAM:        4864 B       768 KB      0.62%
           PSRAM:       2000 KB         8 MB     24.41%
        IDT_LIST:          0 GB        32 KB      0.00%
```

**b_u585i_iot02a:**
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       34680 B         2 MB      1.65%
             RAM:        6960 B       768 KB      0.89%
           PSRAM:       8000 KB         8 MB     97.66%
          EXTMEM:          0 GB        64 MB      0.00%
        IDT_LIST:          0 GB        32 KB      0.00%
```

```
Hello STM32 OSPI PSRAM!
Verifying PSRAM Buffer size: 2048000 ...
PSRAM verification successful!
[00:00:00.000,000] <dbg> memc_stm32_ospi_psram: memc_stm32_ospi_psram_init: OSPI AHB clock frequency: 160000000 Hz
[00:00:00.000,000] <dbg> memc_stm32_ospi_psram: memc_stm32_ospi_psram_init: OSPI max frequency: 133000000 Hz
[00:00:00.000,000] <dbg> memc_stm32_ospi_psram: memc_stm32_ospi_psram_init: clk: 80000000, prescaler: 2
[00:00:00.000,000] <dbg> memc_stm32_ospi_psram: memc_stm32_ospi_psram_init: MSB set: 23
*** Booting Zephyr OS build v4.2.0-rc1-130-gd97aff3f8f32 ***
```

## Test application 2
`tests/drivers/memc/ram`

```
Memory region         Used Size  Region Size  %age Used
           FLASH:       36384 B         2 MB      1.73%
             RAM:        5984 B       768 KB      0.76%
           PSRAM:          8 MB         8 MB    100.00%
          EXTMEM:          0 GB        64 MB      0.00%
        IDT_LIST:          0 GB        32 KB      0.00%
```

```
*** Booting Zephyr OS build v4.2.0-3146-gea283133abd3 ***
Running TESTSUITE test_ram
===================================================================
START - test_psram
 PASS - test_psram in 0.138 seconds
===================================================================
START - test_ram0
 SKIP - test_ram0 in 0.001 seconds
===================================================================
START - test_sdram1
 SKIP - test_sdram1 in 0.001 seconds
===================================================================
START - test_sdram2
 SKIP - test_sdram2 in 0.001 seconds
===================================================================
START - test_sram1
 SKIP - test_sram1 in 0.001 seconds
===================================================================
START - test_sram2
 SKIP - test_sram2 in 0.001 seconds
===================================================================
TESTSUITE test_ram succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [test_ram]: pass = 1, fail = 0, skip = 5, total = 6 duration = 0.143 seconds
 - PASS - [test_ram.test_psram] duration = 0.138 seconds
 - SKIP - [test_ram.test_ram0] duration = 0.001 seconds
 - SKIP - [test_ram.test_sdram1] duration = 0.001 seconds
 - SKIP - [test_ram.test_sdram2] duration = 0.001 seconds
 - SKIP - [test_ram.test_sram1] duration = 0.001 seconds
 - SKIP - [test_ram.test_sram2] duration = 0.001 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

mariopaja commented:
> Thank you Mario! Could you tag this for backport into v4.1?

@dkouba-atym I dont think I can or am allowed to. I dont really see any relevance of this driver for v4.1.

JarmouniA reviewed (commented):
Needs to be enabled on at least one board in tests.

Ductapemaster commented:
> > Hey team, we've rebased on 4.2 so no need to backport this anymore.
> > I'd like to nudge this for review — it's been working great for us.
> 
> @dkouba-atym this would take a bit more time from my side since I am more focused on stm32 sai at the moment

Understood, and I appreciate your efforts!

mariopaja commented on drivers/memc/memc_stm32_ospi_psram.c:
```diff
@@ -0,0 +1,456 @@
+	LOG_DBG("OSPI AHB clock frequency: %d Hz", ahb_clock_freq);
+	LOG_DBG("OSPI max frequency: %d Hz", dev_cfg->max_frequency);
+
+	for (; prescaler <= STM32_OSPI_CLOCK_PRESCALER_MAX; prescaler++) {
```
@etienne-lms 

you mean instead of having it at the beginig of the function `uint32_t prescaler = STM32_OSPI_CLOCK_PRESCALER_MIN;` to declare only `uint32_t prescaler;` and then  set the prescaler value just before the loop?

I did it so to follow the same logic as in `xspi_psram`

etienne-lms replied:
I think it would be more readable, but that not a strong opinion.

mariopaja replied:
done :)

mariopaja commented on dts/bindings/memory-controllers/st,stm32-ospi-psram.yaml:
```diff
@@ -0,0 +1,143 @@
+  fixed-latency:
+    type: boolean
+    description: |
+      Read latency type: variable if not set, fixed if set.
```
Set to single line :)

maass-hamburg commented on drivers/ethernet/Kconfig.oa_tc6:
```diff
@@ -0,0 +1,47 @@
+# Copyright (c) 2025 Microchip Technology Inc.
```
this file seems to not belong in this commit and pr

mariopaja commented on dts/bindings/memory-controllers/st,stm32-ospi-psram.yaml:
```diff
@@ -0,0 +1,144 @@
+    type: int
+    default: 2
+    description: |
+      Write latency code:
```
write-latency might be wrong

mariopaja commented:
@etienne-lms 
> could you add a DTS overlay file tests/drivers/memc/ram/boards/b_u585i_iot02a.overlay

sure :)

> Would it be worth to add your test sample https://github.com/mariopaja/hello_psram in the zephyr tree for testing over bigger buffers (a few MByte) as often found with PSRAM, intead of the 64 byte only of the memc/ram driver test?
Or alternatively, maybe test/drivers/memc/ram test could be extended to have the test buffer size configurable so we can exercice the read/write test over some big buffers.

If it is needed I can push the sample to zephyr tree. I just need to know in which directory to push it and the name. This would be the easiest for me.

Otherwise I could have a look into `test/drivers/memc/ram` :)

etienne-lms reviewed (commented):
Could you add a short sentence in the commit message like the below?
```
Add an overlay DTSI in tests/drivers/memc/ram for that board to pass this test
over its PSRAM only.
```

maass-hamburg commented on dts/bindings/memory-controllers/st,stm32-ospi-psram.yaml:
```diff
@@ -0,0 +1,146 @@
+# Copyright (c) 2025 ZAL Zentrum für Angewandte Luftfahrtforschung GmbH
+# Copyright (c) 2025 Mario Paja
```
either you or the company you work for, also take a look at https://docs.zephyrproject.org/latest/contribute/guidelines.html#copyrights-notices

mariopaja replied:
@maass-hamburg I understand, but it is a work that is on the clock & on private time. I could remove it from the .yaml but I would not from the driver

Ductapemaster commented:
Is there a reason why we are explicitly naming the linker regions in the `.ld` file rather than letting the device tree label define it? Other boards let me use the label provided directly.

Piziwate commented:
> Hi @Piziwate, I dont have the HW. I cannot do tests or provide input :/
> 
> BTW is uses 128-Mbit HyperRAM™ based on [STM32H735G-DK ](https://www.st.com/en/evaluation-tools/stm32h735g-dk.html)

Yes, I know—and that’s exactly my problem. This MCU is interesting for our project, and using HyperRAM makes for a simple design. Unfortunately, I’m not quite sure how to integrate it with Zephyr. I thought Octo PSRAM was similar. I tried using the driver from this PR, but I’m getting errors like: ‘error: unknown type name ‘HAL_OSPI_DLYB_CfgTypeDef’.’
I'm happy to run tests, but I have to admit I don't know the STM32 HAL well enough to develop a new driver.

mariopaja commented:
> Yes, I know—and that’s exactly my problem. This MCU is interesting for our project, and using HyperRAM makes for a simple design. Unfortunately, I’m not quite sure how to integrate it with Zephyr. I thought Octo PSRAM was similar. I tried using the driver from this PR, but I’m getting errors like: ‘error: unknown type name ‘HAL_OSPI_DLYB_CfgTypeDef’.’
> I'm happy to run tests, but I have to admit I don't know the STM32 HAL well enough to develop a new driver.

I think `HAL_OSPI_DLYB_CfgTypeDef` is U5xx specific

erwango commented on drivers/memc/memc_stm32_ospi_psram.c:
```diff
@@ -0,0 +1,460 @@
+	struct memc_stm32_ospi_psram_data *dev_data = dev->data;
+	OSPI_HandleTypeDef *hospi = &dev_data->hospi;
+	OSPI_RegularCmdTypeDef sCommand = {0};
+	OSPI_MemoryMappedTypeDef sMemMappedCfg = {0};
```
```suggestion
	OSPI_MemoryMappedTypeDef mem_mapped_cfg = {0};
```

Piziwate commented:
> > Yes, I know—and that’s exactly my problem. This MCU is interesting for our project, and using HyperRAM makes for a simple design. Unfortunately, I’m not quite sure how to integrate it with Zephyr. I thought Octo PSRAM was similar. I tried using the driver from this PR, but I’m getting errors like: ‘error: unknown type name ‘HAL_OSPI_DLYB_CfgTypeDef’.’
> > I'm happy to run tests, but I have to admit I don't know the STM32 HAL well enough to develop a new driver.
> 
> I think `HAL_OSPI_DLYB_CfgTypeDef` is U5xx specific

Okay, that makes more sense! In that case, shouldn’t you add a dependency at the Kconfig level to prevent it from being (unsuccessfully) used on other STM32s?

erwango commented on drivers/memc/Kconfig.stm32:
```diff
@@ -40,3 +41,15 @@ config MEMC_STM32_XSPI_PSRAM
+	depends on DT_HAS_ST_STM32_OSPI_PSRAM_ENABLED
+	select USE_STM32_HAL_OSPI
+	select USE_STM32_LL_DLYB
+	select USE_STM32_HAL_DMA_EX
+	select USE_STM32_HAL_DMA
```
Are these required by OSPI HAL ?
Otherwise not required

mariopaja replied:
The dependency comes from `HAL_OSPI_Abort`

```
/Users/mariopaja/ZephyrSDK/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: 

modules/hal_stm32/stm32cube/lib..__modules__hal__stm32__stm32cube.a(stm32u5xx_hal_ospi.c.obj): 
in function `HAL_OSPI_Abort':

/Users/mariopaja/zephyrproject/modules/hal/stm32/stm32cube/stm32u5xx/drivers/src/stm32u5xx_hal_ospi.c:2391: 

undefined reference to `HAL_DMA_Abort'
```

erwango reviewed (changes requested):
Thanks for pushing this !

mariopaja commented:
> Okay, that makes more sense! In that case, shouldn’t you add a dependency at the Kconfig level to prevent it from being (unsuccessfully) used on other STM32s?

@JarmouniA @erwango should I make it U5xx specific, or should I do the following:

> As a naive fix, that's unlikely to work correctly, try guarding HAL_OSPI_DLYB....  calls with #ifdef CONFIG_SOC_SERIES_STM32U5X... #endif and modify driver config to select USE_STM32_LL_DLYB if SOC_SERIES_STM32U5X

etienne-lms commented on dts/bindings/memory-controllers/st,stm32-ospi-psram.yaml:
```diff
@@ -0,0 +1,149 @@
+      - 4
+      - 5
+
+  refresh-rate:
```
The property name is a bit ambiguous. How about `adaptive-refresh-rate` or `auto-adjust-refresh-rate`?

etienne-lms commented on dts/bindings/memory-controllers/st,stm32-ospi-psram.yaml:
```diff
@@ -0,0 +1,149 @@
+    type: int
+    default: 0
+    description: |
+      Refresh coverage:
```
For readability (of `pasr` property name acronym), I would suggest `Partial Array Self-Refresh coverage:`

etienne-lms reviewed (commented):
I think there should be 1 commit for the driver part (incl. the DT bindings addition), 1 for board b_u585i_iot02a and a last one to enable test on that board, which may request to add `memc` tag in the `supported` list in the board YAML file. Maybe these 2 last commits could be squashed together.

mariopaja committed:
drivers: memc: add driver for stm32 ospi psram

Add a driver for STM32U5 OSPI PSRAM in memory mapped mode.

Signed-off-by: Mario Paja <mariopaja@hotmail.com>

mariopaja committed:
boards: st: b_u585i_iot02a: enable ospi psram node

Changes enable OSPI PSRAM on b_u585i_iot02a board.

Signed-off-by: Mario Paja <mariopaja@hotmail.com>

mariopaja committed:
tests: drivers: memc: add bigger buffer size for psram nodes

This change updates buffer size to the full psram size when
psram node is available

Signed-off-by: Mario Paja <mariopaja@hotmail.com>

mariopaja committed:
tests: drivers: memc: add b_u585i_iot02a

Add b_u585i_iot02a board to test/drivers/memc/ram

Signed-off-by: Mario Paja <mariopaja@hotmail.com>

mariopaja commented on tests/drivers/memc/ram/src/main.c:
```diff
@@ -10,7 +11,12 @@
 
 /** Buffer size. */
+#if DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(psram))
+#define BUF_SIZE (DT_REG_SIZE(DT_NODELABEL(psram)) / 4)
```
What buf_size do you suggest?

etienne-lms reviewed (commented):
Some late comments.

dsseng reviewed (commented):
Some nitpicks

mariopaja commented on tests/drivers/memc/ram/src/main.c:
```diff
@@ -10,7 +11,12 @@
 
 /** Buffer size. */
+#if DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(psram))
+#define BUF_SIZE 524288U
```
Before it was set to `#define BUF_SIZE (DT_REG_SIZE(DT_NODELABEL(psram)) / 4)` but https://github.com/zephyrproject-rtos/zephyr/pull/92381#discussion_r2372338349
